### PR TITLE
generic proxy: dubbo codec for generic proxy

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: generic proxy: dubbo codec for generic proxy

Additional Description:

Add dubbo codec for generic proxy to support proxy dubbo traffic by the generic proxy. It's simple wrapper to the common dubbo codec (`/source/extensions/common/dubbo`).

Risk Level: n/a. new feature.
Testing: Unit.
Docs Changes: n/a.
Release Notes: Added.
Platform Specific Features: n/a.

### CODE REVIEW

File: contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  case StatusCode::kOk:
+    return Common::Dubbo::ResponseStatus::Ok;
+  case StatusCode::kInvalidArgument:
+    return Common::Dubbo::ResponseStatus::BadRequest;
+  default:
+    return Common::Dubbo::ResponseStatus::ServerError;
+  }
+}
+
+StatusCode statusToGenericStatus(Common::Dubbo::ResponseStatus status) {
```
Comment (kyessenov) [inline contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc:51]: These two functions don't seem to commute. Should they somehow map statuses one-to-one?

File: contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  }
+}
+
+} // namespace
+
+void DubboRequest::forEach(IterateCallback callback) const {
+  ASSERT(dynamic_cast<Common::Dubbo::RpcRequestImpl*>(&inner_metadata_->mutableRequest()) !=
+         nullptr);
+  const auto* typed_request =
+      dynamic_cast<Common::Dubbo::RpcRequestImpl*>(&inner_metadata_->mutableRequest());
```
Comment (kyessenov) [inline contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc:79]: Suggestion: `ASSERT(typed_request != nullptr)`.

**Review by kyessenov — COMMENTED** [review]
I'm probably not the best person to review this since I don't now dubbo, but happy to provide some comments.

**Comment by wbpcode** [comment]
@kyessenov this is not particularly relevant to dubbo. It's just a wrapper. :) And take it easy, this is `contrib` extension which is under active development. Some more following PRs are necessary to refine it.

File: contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc (hunk 2 above)
Comment (wbpcode) [reply contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc:79]: got it.

File: contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc (hunk 1 above)
Comment (wbpcode) [reply contrib/generic_proxy/filters/network/source/codecs/dubbo/config.cc:51]: there is no way to map Status (alias of absl::Status) and ResponseStatus (defined by dubbo protocol) one-to-one.
this is an expected flaw.

**Comment by wbpcode** [comment]
/wait

Waiting for #23418 and `extensions/common/dubbo` to be updated first.

**Comment by wbpcode** [comment]
Hi, @lizan, could you also take a look for this pr when you have free time? This pr create a dubbo codec for the generic_proxy (contrib extension) to make the generic_proxy could be used to proxy dubbo traffic.

This is a wrapper/adapter based on the `extensions/common/dubbo`.

File: contrib/contrib_build_config.bzl [hunk 3]
```diff
@@ -59,6 +59,12 @@ CONTRIB_EXTENSIONS = {
 
     "envoy.regex_engines.hyperscan":                            "//contrib/hyperscan/regex_engines/source:config",
 
+    #
+    # Extensions for generic proxy
+    #
+    "envoy.filters.generic.router":                             "//contrib/generic_proxy/filters/network/source/router:config",
+    "envoy.generic_proxy.codecs.dubbo":                          "//contrib/generic_proxy/filters/network/source/codecs/dubbo:config",
```
Comment (lizan) [inline contrib/contrib_build_config.bzl:66]: style nit: align

File: contrib/generic_proxy/filters/network/test/codecs/dubbo/config_test.cc [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+inline void addInt32(Buffer::Instance& buffer, uint32_t value) {
+  value = htobe32(value);
+  buffer.add(&value, 4);
+}
+
+inline void addInt64(Buffer::Instance& buffer, uint64_t value) {
+  value = htobe64(value);
+  buffer.add(&value, 8);
+}
```
Comment (lizan) [inline contrib/generic_proxy/filters/network/test/codecs/dubbo/config_test.cc:34]: User Buffer::Instance has writeBEInt so you can just use them directly.

File: contrib/contrib_build_config.bzl (hunk 3 above)
Comment (wbpcode) [reply contrib/contrib_build_config.bzl:66]: get it.

File: contrib/generic_proxy/filters/network/test/codecs/dubbo/config_test.cc (hunk 4 above)
Comment (wbpcode) [reply contrib/generic_proxy/filters/network/test/codecs/dubbo/config_test.cc:34]: get it.

**Comment by lizan** [comment]
can you resolve conflicts (as result of 1.24.0 release)

**Comment by wbpcode** [comment]
got it

**Review by lizan — APPROVED** [review]

**Comment by wbpcode** [comment]
resolved new conflicts. could you give a lgtm again 🤣 @lizan

**Review by lizan — APPROVED** [review]

**Pull request merged**